docs/enhancements/on-prem-data-collection.md

```diff
@@ -0,0 +1,121 @@
+---
+title: on-prem-data-collection
+authors:
+  - "@CrystalChun"
+creation-date: 2022-09-30
+last-updated: 2022-09-30
+---
+
+# On-premises Data Collection
+
+## Summary
+
+To provide a better installation process and experience, the Assisted Installer ("AI") will collect data about installations when it is deployed on-premises as part of either [Red Hat Advanced Cluster Management for Kubernetes](https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.6) ("RHACM") or as a stand-alone Podman deployment. Data was already collected from clusters deployed on [Red Hat's Hybrid Cloud Console](https://console.redhat.com/openshift/assisted-installer/clusters) (“SaaS”). The collection of data from on-premises deployments will give the team additional insight into the product's usage, customers' pain points, and analytics for how to best improve the product.
+
+## Motivation
+
+Data helps us see how the product is being used, helps us to analyze trends, and to determine ways we can improve the product for our users. 
+
+### Goals
+
+1. Create a process to collect and send [event](https://github.com/openshift/assisted-service/tree/master/docs/dev/events.md) data from on-premises deployments of AI.
+2. Analyze and process on-premises data (by use of visualizations, dashboards, etc.) at Red Hat like SaaS data.
+
+
+### Non-Goals
+
+1. Implementing new data types to collect is out of the scope of this epic. This epic defines how and when the data already collected will be processed and sent.
+2. Define how SaaS should redesign their data pipeline (taken care of in [MGMT-11244](https://issues.redhat.com/browse/MGMT-11244)) is out of the scope of this epic. This epic only refers to on-premises deployments of AI.
+3. Define the creation and/or management of an event streaming instance such as Kafka and the endpoint this data will be hosted at (both taken care of in [MGMT-11244](https://issues.redhat.com/browse/MGMT-11244)).
+4. Collecting must-gather data (and logs) for debugging failed cluster installation is outside the scope of this epic. This epic is only concerned with [event](https://github.com/openshift/assisted-service/blob/6b8401d3b0f0ceea5a837f33c17b9b1cd5ec52a3/internal/events/api/event.go) data.
+
+
+## Proposal
+
+A new flow for sending data is to be added:
+When a cluster installation attempt finishes (regardless of success or failure), all of the event data for this cluster will be queried from the DB, packaged, and sent to the data hosting endpoint.
```

Comment:
We will loose "deserted installations" insights. Is this acceptable?

Reply:
Sorry, could you clarify "deserted installations"? Does that mean that the customer created a cluster, but never starts trying to install it?

Reply:
Yes, something along these lines.
Our funnel goes like this:
1) Cluster creation
2) Adding host
3) Installing cluster
4) Installed

With the current proposal we are only covering point 4, but all the others will be uncovered.
By "deserted installations" we mean clusters that reach stage 2 but never stage 3, but I've just realized that we are also missing other cases.

Reply:
I see what you mean. Do we currently collect data at all stages mentioned in SaaS now? Also would the data that hasn't surpassed stage 3 be relevant or interesting?

Reply:
Yes, at the moment in SaaS we collect all events (not really but that's the intention).
We are currently analyzing "deserted installations" (Reached stage 2 but not stage 3): these users are users that did interact with our application (not just creating a cluster, but bothered adding a host), but for some reason could not reach the installation stage.
If we can detect what are the reasons, in some cases we could help the users go through these hurdles (by enhancing or adding features or simply improving the UI).

I wouldn't say it's critical but it sure helps.

Reply:
Gotcha, that makes sense! Thanks for explaining!

Trying to think of a better flow to capture this data without overloading the ingress http server, any ideas? If the customer reaches stage 2 but not stage 3, would they delete the cluster or start removing hosts? If so we could probably trigger a data send when they delete/remove?

Reply:
> Trying to think of a better flow to capture this data without overloading the ingress http server, any ideas? If the customer reaches stage 2 but not stage 3, would they delete the cluster or start removing hosts? If so we could probably trigger a data send when they delete/remove?

In some cases, yes. I think that a kind of shutdown trigger that would send all collected that it's a very good idea, although it won't work in the "unplugging the machine" situation but at least we cover a lot of cases.

Maybe we can also have a periodic check (long interval, maybe daily or something?) and if the cluster has not changed state (meaning the installation is abandoned/forgotten), just send the data we have until that point? WDYT?

Reply:
FYI Rom suggested for these to be future enhancements